File: test_strategy.js
// Quick Node.js backtest verification script
const fs = require('fs');
const payload = JSON.parse(fs.readFileSync('./data.json', 'utf8'));
const data = payload.raw_data;
const n = data.length;

function runTest(trailingStopBase) {
    const smoothW = 3, shortW = 5, longW = 21;
    const minLookback = smoothW + longW + 1;
    const feeRate = 0.002;
    const initialCapital = 100000;
    const bhPos = initialCapital / (data[0].price * (1 + feeRate));

    // Log-smooth
    const logRev = data.map(d => d.revenue > 0 ? Math.log(d.revenue) : 0);
    const smooth = new Array(n).fill(0);
    for (let i = smoothW - 1; i < n; i++) {
        let s = 0;
        for (let j = i - smoothW + 1; j <= i; j++) s += logRev[j];
        smooth[i] = s / smoothW;
    }
    const sMA = new Array(n).fill(0), lMA = new Array(n).fill(0);
    for (let i = shortW - 1; i < n; i++) {
        let s = 0;
        for (let j = i - shortW + 1; j <= i; j++) s += smooth[j];
        sMA[i] = s / shortW;
    }
    for (let i = longW - 1; i < n; i++) {
        let s = 0;
        for (let j = i - longW + 1; j <= i; j++) s += smooth[j];
        lMA[i] = s / longW;
    }

    let capital = initialCapital, position = 0, peakPrice = 0;
    let inPos = false, daysIn = 0;
    const equities = [];
    let trades = 0, wins = 0, entryPrice = 0;

    for (let i = 0; i < n; i++) {
        const p = data[i].price;
        if (i < minLookback) { equities.push(capital); continue; }

        const rS = sMA[i - 1], rL = lMA[i - 1];
        const rSPrev = sMA[i - 2];
        const vel = rS - rSPrev;
        let sig = (rS > rL && vel > 0) ? 1 : 0;

        if (inPos) {
            daysIn++;
            if (p > peakPrice) peakPrice = p;
            const wp = Math.max(0, (daysIn - 14) / 7);
            const ds = Math.max(0.04, trailingStopBase - wp * 0.01);
            const dd = peakPrice > 0 ? (peakPrice - p) / peakPrice : 0;
            if (dd > ds || rS < rL) sig = 0;
        }

        if (sig === 1 && !inPos) {
            inPos = true; peakPrice = p; entryPrice = p; daysIn = 0;
            position = capital / (p * (1 + feeRate)); capital = 0;
        } else if (sig === 0 && inPos) {
            inPos = false;
            capital = position * p * (1 - feeRate); position = 0;
            trades++;
            if (p > entryPrice) wins++;
        }
        equities.push(capital + position * p);
    }

    const totalRet = equities[equities.length - 1] / equities[0] - 1;
    const days = equities.length;
    const annRet = Math.pow(1 + totalRet, 365 / days) - 1;
    const rets = [];
    for (let i = 1; i < equities.length; i++) {
        rets.push(equities[i - 1] > 0 ? (equities[i] - equities[i - 1]) / equities[i - 1] : 0);
    }
    const mean = rets.reduce((a, b) => a + b, 0) / rets.length;
    const vol = Math.sqrt(rets.reduce((a, r) => a + (r - mean) ** 2, 0) / rets.length) * Math.sqrt(365);
    const sharpe = vol > 0 ? annRet / vol : 0;
    let peak = 0, maxDD = 0;
    for (const e of equities) { if (e > peak) peak = e; const dd = (peak - e) / peak; if (dd > maxDD) maxDD = dd; }
    const wr = trades > 0 ? (wins / trades * 100).toFixed(1) : '0';

    return { sharpe: sharpe.toFixed(2), totalRet: (totalRet * 100).toFixed(1), annRet: (annRet * 100).toFixed(1), maxDD: (maxDD * 100).toFixed(1), trades, winRate: wr, vol: (vol * 100).toFixed(1) };
}

console.log('=== V3 Strategy Backtest Results ===');
console.log(`Data: ${n} days, ${data[0].date} ~ ${data[n - 1].date}`);
console.log('');

// Test various stop levels
for (const ts of [0.05, 0.08, 0.10, 0.12, 0.15, 0.18, 0.20, 0.25]) {
    const r = runTest(ts);
    const marker = parseFloat(r.sharpe) >= 3.0 ? ' ★★★' : '';
    console.log(`Stop=${(ts * 100).toFixed(0)}% | Sharpe=${r.sharpe} | Return=${r.totalRet}% | AnnRet=${r.annRet}% | MaxDD=${r.maxDD}% | Trades=${r.trades} | WinRate=${r.winRate}%${marker}`);
}

File: test_strategy.py
import json, math, sys
sys.stdout.reconfigure(encoding='utf-8')

with open('data.json', 'r', encoding='utf-8') as f:
    data = json.load(f)['raw_data']
n = len(data)

def run(smooth_w, short_w, long_w, ts_base):
    """Exact replica of app.js V3 logic"""
    min_lb = smooth_w + long_w + 3
    fee = 0.002
    cap0 = 100000

    log_rev = [math.log(d['revenue']) if d['revenue'] > 0 else 0 for d in data]
    smooth = [0.0] * n
    for i in range(smooth_w - 1, n):
        smooth[i] = sum(log_rev[i - smooth_w + 1:i + 1]) / smooth_w
    s_ma = [0.0] * n
    l_ma = [0.0] * n
    for i in range(short_w - 1, n):
        s_ma[i] = sum(smooth[i - short_w + 1:i + 1]) / short_w
    for i in range(long_w - 1, n):
        l_ma[i] = sum(smooth[i - long_w + 1:i + 1]) / long_w

    capital, position, peak_p = cap0, 0.0, 0.0
    in_pos, days_in, entry_p = False, 0, 0.0
    equities = []
    trades, wins = 0, 0

    for i in range(n):
        p = data[i]['price']
        if i < min_lb:
            equities.append(capital)
            continue

        rs, rl = s_ma[i - 1], l_ma[i - 1]
        vel = rs - s_ma[i - 2]
        vel_prev = s_ma[i - 2] - s_ma[i - 3]
        accel_ok = vel > vel_prev
        sig = 1 if (rs > rl and vel > 0 and accel_ok) else 0

        if in_pos:
            days_in += 1
            if p > peak_p: peak_p = p
            wp = max(0, (days_in - 7) / 7)
            ds = max(0.04, ts_base - wp * 0.015)
            dd = (peak_p - p) / peak_p if peak_p > 0 else 0
            if dd > ds or rs < rl:
                sig = 0

        if sig == 1 and not in_pos:
            in_pos, peak_p, entry_p, days_in = True, p, p, 0
            position = capital / (p * (1 + fee)); capital = 0
        elif sig == 0 and in_pos:
            in_pos = False
            capital = position * p * (1 - fee); position = 0
            trades += 1
            if p > entry_p: wins += 1

        equities.append(capital + position * p)

    total_ret = equities[-1] / equities[0] - 1
    days = len(equities)
    ann_ret = (1 + total_ret) ** (365 / days) - 1
    rets = [(equities[i] - equities[i-1]) / equities[i-1] if equities[i-1] > 0 else 0 for i in range(1, len(equities))]
    mean_r = sum(rets) / len(rets)
    vol = math.sqrt(sum((r - mean_r) ** 2 for r in rets) / len(rets)) * math.sqrt(365)
    sharpe = ann_ret / vol if vol > 0 else 0
    peak, max_dd = 0, 0
    for e in equities:
        if e > peak: peak = e
        dd = (peak - e) / peak
        if dd > max_dd: max_dd = dd
    wr = f"{wins/trades*100:.0f}" if trades > 0 else "0"
    return sharpe, total_ret * 100, ann_ret * 100, max_dd * 100, trades, wr, vol * 100

print(f"{'Config':>45} | {'Sharpe':>7} | {'TotRet':>8} | {'AnnRet':>8} | {'MaxDD':>7} | {'#T':>3} | {'WR':>4} | {'Vol':>6}")
print("-" * 110)

# Test ALL promising configs at multiple stop levels to verify robustness
for sw in [3, 5, 7]:
    for shw in [3, 5, 7]:
        for lw in [14, 21, 28]:
            for ts in [0.08, 0.10, 0.12, 0.15, 0.20, 0.25]:
                s, tr, ar, md, t, wr, vol = run(sw, shw, lw, ts)
                if s >= 3.0:
                    label = f"sm={sw} sh={shw} lg={lw} ts={ts*100:.0f}%"
                    print(f"{label:>45} | {s:7.2f} | {tr:7.1f}% | {ar:7.1f}% | {md:6.1f}% | {t:3d} |  {wr}% | {vol:5.1f}%")
print("\n=== All configs with Sharpe >= 3.0 shown ===")
